lib/served/resource/http_errors.rb

```diff
@@ -0,0 +1,140 @@
+module Served
+  module Resource
+    class HttpError < Served::Error
+      attr_reader :message
+      attr_reader :server_backtrace
+      attr_reader :errors
+
+      # Defined in individual error classes
+      def self.code
+        raise NotImplementedError
+      end
+
+      def initialize(resource, response)
+        if resource.serializer.respond_to? :exception
+          serialized = resource.serializer.exception(response.body).symbolize_keys!
+
+          @error            = serialized[:error]
+          @message          = serialized[:exception]
+          @server_backtrace = serialized[:backtrace]
+          @code             = serialized[:code]
+
+          super("An error '#{code} #{message}' occurred while making this request")
+        end
+        super "An error occurred '#{self.class.code}'"
+      end
+
+      def status
+        self.class.status
+      end
+
+      def code
+        @code || self.class.code
+      end
+
+    end
+
+    # 301 MovedPermanently
+    class MovedPermanently < HttpError
+
+      def self.code
+        301
+      end
+
+    end
+
+    # 400 BadRequest
+    class BadRequest < HttpError
+
+      def self.code
+        400
+      end
+
+    end
+
+    # 401 Unauthorized
+    class Unauthorized < HttpError
+
+      def self.code
+        401
+      end
+
+    end
+
+    # 403 Forbidden
+    class Forbidden < HttpError
+
+      def self.code
+        403
+      end
+
+    end
+
+    # 404 NotFound
+    class NotFound < HttpError
+
+      def self.code
+        404
+      end
+
+    end
+
+    # 405 MethodNotAllowed
+    class MethodNotAllowed < HttpError
+
+      def self.code
+        405
+      end
+
+    end
+
+    # 406 NotAcceptable
+    class NotAcceptable < HttpError
+
+      def self.code
+        406
+      end
+
+    end
+
+    # 408 RequestTimeout
+    class RequestTimeout < HttpError
+
+      def self.code
+        408
+      end
+
+    end
+
+    # 422 UnprocessableEntity
+    class UnprocessableEntity < HttpError
+
+      def self.code
+        422
+      end
+
+    end
+
+    # 500 InternalServerError
+    class InternalServerError < HttpError
+
+      def self.code
+        500
+      end
+
+    end
+
+    # 503 BadGateway
+    class BadGateway < HttpError
```

Comment:
Not a deal breaker, but the sheer number of these subclasses seems verbose.  I'm wondering if it'd be better to have a single Error handling class with behavior and then the various specific errors would just be data in the form of structs or something.